### PR TITLE
Add mod_gil_not_used call to PYBIND11_MODULE

### PR DESCRIPTION
**Pull request by lysnikolaou**
Related to #407.

It'd probably be best to merge #408 first, so that this gets CI coverage on the free-threaded build.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@ build-backend = "mesonpy"
 requires = [
     "meson >= 1.2.0",
     "meson-python >= 0.13.1",
-    "pybind11 >= 2.12.0",
+    "pybind11 >= 2.13.0",
```
Comment (rgommers) [inline pyproject.toml:6]: 2.13.1 was released the day after 2.13.0 ((IIRC with a free-threading related fix as well), so I'd always use .1 as the minimum version

**Comment by lysnikolaou** [comment]
This is working correctly. Examining [the test logs](https://github.com/contourpy/contourpy/actions/runs/10078841335/job/27864748623#step:17:3723) shows that there's no warnings that the GIL is re-enabled during the tests.

**Comment by lysnikolaou** [comment]
Cirrus failure is unrelated.

**Review by ianthomas23 — APPROVED** [review]
Thanks @lysnikolaou. I was expecting demands for free-threaded support to appear shortly, so I am very happy that someone knowledgeable is helping here.

**Comment by lysnikolaou** [comment]
Glad to be of help @ianthomas23! If there's anything else I can help out with, feel free to ping me. Also thanks for the speedy reviews!

**Pull request merged**